### PR TITLE
Use type on call path, and minor refactoring

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Now instead of using the call type, we add the type to the path at redux level.

I also included here the AppCall/AppCallRequest needed changes.

The errors in CI are most probably due to redux PR not being merged. When the Redux PR is merged, I will update this PR dependencies.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33178

#### Related Pull Requests
Redux: https://github.com/mattermost/mattermost-redux/pull/1406
Mobile: https://github.com/mattermost/mattermost-mobile/pull/5221
Proxy: https://github.com/mattermost/mattermost-plugin-apps/pull/99

### CODE REVIEW

File: components/post_view/embedded_bindings/select_binding/select_binding.tsx [hunk 1]
```diff
@@ -72,19 +74,29 @@ class SelectBinding extends React.PureComponent<Props, State> {
...
 
-        const {userId, post} = this.props;
+        const {post} = this.props;
 
-        const call = createCallRequest(
-            binding.call,
-            userId,
+        let teamID = '';
+        const {data} = await this.props.actions.getChannel(post.channel_id) as {data?: any; error?: any};
+        if (data) {
```
Comment (catalintomai) [inline components/post_view/embedded_bindings/select_binding/select_binding.tsx:81]: `if (data) -> if(data & data is Channel)` ? (ensures the `data as Channel` below is successful)

File: components/post_view/embedded_bindings/button_binding/button_binding.tsx [hunk 2]
```diff
@@ -3,25 +3,27 @@
...
 
 type Props = {
     intl: IntlShape;
     binding: AppBinding;
-    userId: string;
     post: Post;
     actions: {
-        doAppCall: (call: AppCall) => Promise<ActionResult>;
+        doAppCall: (call: AppCallRequest, type: AppCallType) => Promise<ActionResult>;
+        getChannel: (channelId: string) => Promise<ActionResult>;
```
Comment (catalintomai) [inline components/post_view/embedded_bindings/button_binding/button_binding.tsx:26]: is this part of the "minor refactoring" you refer to in the PR description?

Comment (larkox) [reply components/post_view/embedded_bindings/button_binding/button_binding.tsx:26]: I have mixed feelings with this. What we were doing was getting the current team id. What we are doing now, is getting the channel where the post is to get the team id of the post.

At the moment, you cannot be in a thread of a different team. But if that was to change at any moment, this approach is more robust.

Nevertheless, I can go back to the previous approach.

**Review by levb — APPROVED** [review]

**Review by catalintomai — APPROVED** [review]

File: components/dot_menu/dot_menu.tsx [hunk 3]
```diff
@@ -283,18 +284,22 @@ class DotMenu extends React.PureComponent<Props, State> {
...
+        );
+        const call = createCallRequest(
+            binding.call,
+            context,
+            {
+                post: AppExpandLevels.ALL,
             },
-        });
+        );
+        const res = await this.props.actions?.doAppCall(call, AppCallTypes.SUBMIT);
```
Comment (mickmister) [inline components/dot_menu/dot_menu.tsx:302]: Why is actions optional?

File: components/post_view/embedded_bindings/button_binding/button_binding.tsx [hunk 4]
```diff
@@ -3,25 +3,27 @@
...
-import {AppBinding, AppCall, AppCallResponse} from 'mattermost-redux/types/apps';
+import {AppBinding, AppCallRequest, AppCallResponse, AppCallType} from 'mattermost-redux/types/apps';
 import {ActionResult} from 'mattermost-redux/types/actions';
-import {AppBindingLocations, AppCallResponseTypes, AppExpandLevels} from 'mattermost-redux/constants/apps';
+import {AppBindingLocations, AppCallResponseTypes, AppCallTypes, AppExpandLevels} from 'mattermost-redux/constants/apps';
 import {Post} from 'mattermost-redux/types/posts';
 
 import {injectIntl, IntlShape} from 'react-intl';
 
+import {Channel} from 'mattermost-redux/types/channels';
```
Comment (mickmister) [inline components/post_view/embedded_bindings/button_binding/button_binding.tsx:13]: Can we put this import with the other mm-redux imports?

File: components/post_view/embedded_bindings/button_binding/button_binding.tsx [hunk 5]
```diff
@@ -38,22 +40,32 @@ class ButtonBinding extends React.PureComponent<Props, State> {
...
+        if (data) {
+            const channel = data as Channel;
+            teamID = channel.team_id;
+        }
+
+        const context = createCallContext(
             binding.app_id,
-            AppBindingLocations.IN_POST + '/' + binding.location,
-            {post: AppExpandLevels.EXPAND_ALL},
+            AppBindingLocations.IN_POST + binding.location,
```
Comment (mickmister) [inline components/post_view/embedded_bindings/button_binding/button_binding.tsx:57]: Is it implied there is a slash on `binding.location`?

File: components/post_view/embedded_bindings/select_binding/select_binding.tsx [hunk 6]
```diff
@@ -72,19 +74,29 @@ class SelectBinding extends React.PureComponent<Props, State> {
...
+            const channel = data as Channel;
+            teamID = channel.team_id;
+        }
+
+        const context = createCallContext(
             binding.app_id,
-            AppBindingLocations.IN_POST + '/' + binding.location,
-            {post: AppExpandLevels.EXPAND_ALL},
+            AppBindingLocations.IN_POST + binding.location,
```
Comment (mickmister) [inline components/post_view/embedded_bindings/select_binding/select_binding.tsx:88]: Again is it implied there is a slash on `binding.location`?

**Review by mickmister — CHANGES_REQUESTED** [review]

File: components/dot_menu/dot_menu.tsx (hunk 3 above)
Comment (larkox) [reply components/dot_menu/dot_menu.tsx:302]: Probably the original proptypes did not have the isrequired, and was just migrated like that.

**Review by mickmister — APPROVED** [review]

**Pull request merged**